midterm/ProductDetails.tsx

```diff
@@ -0,0 +1,218 @@
+import { useNavigation, useRoute } from "@react-navigation/native";
+import { useEffect, useRef, useState } from "react";
+import { Text, View } from "react-native";
+import { TouchableOpacity, StyleSheet, Image } from "react-native";
+import Ionicons from "@expo/vector-icons/Ionicons";
+import { useAsyncStorage } from "@react-native-async-storage/async-storage";
+import { useFetch } from "./hooks/useFetch";
+import { SafeAreaView } from "react-native-safe-area-context";
+import { favoriteItem } from "./types/favoriteItem";
+
+type RouteParams = {
+	url: string;
+};
+
+export const ProductDetails = () => {
+	const buttonPressed = useRef(false);
+	const { params }: any = useRoute();
+	const typedParams: RouteParams = params;
+
+	const [favoritesList, setFavoritesList] = useState<favoriteItem[]>([]);
```

Comment:
You'll only need `inFavorites` to be part of state. The flow would look like this:

1. Scan a code and arrive at ProductDetails with a URL
2. Check if this URL is in our async storage and set the state of `inFavorites`
3. Load data from GET request

When interacting with the heart button again:
1. Get async storage
2. Search for matching URL
3. Set favorite or unfavorite
4. Save updated async storage

In either of those scenarios, you can rely on retrieving the list from async storage and not from state and only tapping the favorite button would cause a re-render.